pyproject.toml

```diff
@@ -22,3 +22,19 @@ Homepage = "https://github.com/wjakob/nanobind_example"
 [tool.scikit-build]
 # Build stable ABI wheels for CPython 3.12+
 wheel.py-api = "cp312"
+
+[tool.cibuildwheel]
+# Only build for PyPy 3.9
+skip = ["pp37*", "pp38*"]
```

Comment:
I think that it might work. I was initially focusing on porting nanobind on the latest version, but the needed features should have been ported to the other PyPy Python release branches as well.

Reply:
I'd not be surprised if pypy37 doesn't work, since it stopped being released 1-2 releases ago. But I think it's totally fine to skip that.

Reply:
Nope:

```
[1/13] Building CXX object CMakeFiles/nanobind_example_ext.dir/src/nanobind_example_ext.cpp.o
[2/13] Building CXX object CMakeFiles/nanobind-static.dir/tmp/pip-build-env-k7cixrlv/overlay/lib/pypy3.8/site-packages/nanobind/src/nb_internals.cpp.o
[3/13] Building CXX object CMakeFiles/nanobind-static.dir/tmp/pip-build-env-k7cixrlv/overlay/lib/pypy3.8/site-packages/nanobind/src/nb_func.cpp.o
[4/13] Building CXX object CMakeFiles/nanobind-static.dir/tmp/pip-build-env-k7cixrlv/overlay/lib/pypy3.8/site-packages/nanobind/src/nb_enum.cpp.o
[5/13] Building CXX object CMakeFiles/nanobind-static.dir/tmp/pip-build-env-k7cixrlv/overlay/lib/pypy3.8/site-packages/nanobind/src/nb_type.cpp.o
[6/13] Building CXX object CMakeFiles/nanobind-static.dir/tmp/pip-build-env-k7cixrlv/overlay/lib/pypy3.8/site-packages/nanobind/src/nb_static_property.cpp.o
[7/13] Building CXX object CMakeFiles/nanobind-static.dir/tmp/pip-build-env-k7cixrlv/overlay/lib/pypy3.8/site-packages/nanobind/src/nb_ndarray.cpp.o
[8/13] Building CXX object CMakeFiles/nanobind-static.dir/tmp/pip-build-env-k7cixrlv/overlay/lib/pypy3.8/site-packages/nanobind/src/common.cpp.o
FAILED: CMakeFiles/nanobind-static.dir/tmp/pip-build-env-k7cixrlv/overlay/lib/pypy3.8/site-packages/nanobind/src/common.cpp.o
/opt/rh/devtoolset-10/root/usr/bin/c++ -DNB_COMPACT_ASSERTIONS -I/tmp/pip-build-env-k7cixrlv/overlay/lib/pypy3.8/site-packages/nanobind/ext/robin_map/include -I/opt/python/pp38-pypy38_pp73/include/pypy3.8 -I/tmp/pip-build-env-k7cixrlv/overlay/lib/pypy3.8/site-packages/nanobind/include -O3 -DNDEBUG -std=gnu++17 -fPIC -fvisibility=hidden -ffunction-sections -fdata-sections -fno-strict-aliasing -MD -MT CMakeFiles/nanobind-static.dir/tmp/pip-build-env-k7cixrlv/overlay/lib/pypy3.8/site-packages/nanobind/src/common.cpp.o -MF CMakeFiles/nanobind-static.dir/tmp/pip-build-env-k7cixrlv/overlay/lib/pypy3.8/site-packages/nanobind/src/common.cpp.o.d -o CMakeFiles/nanobind-static.dir/tmp/pip-build-env-k7cixrlv/overlay/lib/pypy3.8/site-packages/nanobind/src/common.cpp.o -c /tmp/pip-build-env-k7cixrlv/overlay/lib/pypy3.8/site-packages/nanobind/src/common.cpp
/tmp/pip-build-env-k7cixrlv/overlay/lib/pypy3.8/site-packages/nanobind/src/common.cpp: In function ‘PyObject* nanobind::detail::obj_vectorcall(PyObject*, PyObject* const*, size_t, PyObject*, bool)’:
/tmp/pip-build-env-k7cixrlv/overlay/lib/pypy3.8/site-packages/nanobind/src/common.cpp:275:51: error: invalid conversion from ‘PyObject* const*’ {aka ‘_object* const*’} to ‘_object**’ [-fpermissive]
  275 |             res = _PyObject_Vectorcall(self, args + 1, nargsf - 1, kwnames);
      |                                              ~~~~~^~~
      |                                                   |
      |                                                   PyObject* const* {aka _object* const*}
In file included from /opt/python/pp38-pypy38_pp73/include/pypy3.8/Python.h:122,
                 from /tmp/pip-build-env-k7cixrlv/overlay/lib/pypy3.8/site-packages/nanobind/include/nanobind/nb_python.h:21,
                 from /tmp/pip-build-env-k7cixrlv/overlay/lib/pypy3.8/site-packages/nanobind/include/nanobind/nanobind.h:38,
                 from /tmp/pip-build-env-k7cixrlv/overlay/lib/pypy3.8/site-packages/nanobind/src/common.cpp:10:
/opt/python/pp38-pypy38_pp73/include/pypy3.8/pypy_decl.h:1139:90: note:   initializing argument 2 of ‘_object* _PyPyObject_Vectorcall(_object*, _object**, Py_ssize_t, _object*)’
 1139 | PyAPI_FUNC(struct _object *) _PyObject_Vectorcall(struct _object *arg0, struct _object **arg1, Signed arg2, struct _object *arg3);
      |                                                                         ~~~~~~~~~~~~~~~~~^~~~
/tmp/pip-build-env-k7cixrlv/overlay/lib/pypy3.8/site-packages/nanobind/src/common.cpp:279:42: error: invalid conversion from ‘PyObject* const*’ {aka ‘_object* const*’} to ‘_object**’ [-fpermissive]
  279 |         res = _PyObject_Vectorcall(base, args, nargsf, kwnames);
      |                                          ^~~~
      |                                          |
      |                                          PyObject* const* {aka _object* const*}
In file included from /opt/python/pp38-pypy38_pp73/include/pypy3.8/Python.h:122,
                 from /tmp/pip-build-env-k7cixrlv/overlay/lib/pypy3.8/site-packages/nanobind/include/nanobind/nb_python.h:21,
                 from /tmp/pip-build-env-k7cixrlv/overlay/lib/pypy3.8/site-packages/nanobind/include/nanobind/nanobind.h:38,
                 from /tmp/pip-build-env-k7cixrlv/overlay/lib/pypy3.8/site-packages/nanobind/src/common.cpp:10:
/opt/python/pp38-pypy38_pp73/include/pypy3.8/pypy_decl.h:1139:90: note:   initializing argument 2 of ‘_object* _PyPyObject_Vectorcall(_object*, _object**, Py_ssize_t, _object*)’
 1139 | PyAPI_FUNC(struct _object *) _PyObject_Vectorcall(struct _object *arg0, struct _object **arg1, Signed arg2, struct _object *arg3);
      |                                                                         ~~~~~~~~~~~~~~~~~^~~~
[9/13] Building CXX object CMakeFiles/nanobind-static.dir/tmp/pip-build-env-k7cixrlv/overlay/lib/pypy3.8/site-packages/nanobind/src/error.cpp.o
[10/13] Building CXX object CMakeFiles/nanobind-static.dir/tmp/pip-build-env-k7cixrlv/overlay/lib/pypy3.8/site-packages/nanobind/src/trampoline.cpp.o
ninja: build stopped: subcommand failed.
```